File: herald/sources.py
from __future__ import annotations

import json
from dataclasses import dataclass
from importlib.resources import files
from typing import Any, Iterable, Mapping
from urllib.parse import urlsplit

from .feeds import canonicalize_url


SOURCE_MANIFEST_FORMAT = "herald.sources"
SOURCE_MANIFEST_VERSION = 1
MAX_SOURCES = 1_000
_SOURCE_FIELDS = {"title", "url", "category", "content_kind", "enabled"}


class SourceManifestError(ValueError):
    """Raised when a portable source manifest is invalid or unsafe."""


@dataclass(frozen=True, slots=True)
class CuratedSource:
    title: str
    url: str
    category: str
    content_kind: str = "paper"
    enabled: bool = True

    def to_dict(self) -> dict[str, str | bool]:
        return {
            "title": self.title,
            "url": self.url,
            "category": self.category,
            "content_kind": self.content_kind,
            "enabled": self.enabled,
        }


def _required_text(item: Mapping[str, Any], key: str, index: int) -> str:
    value = item.get(key)
    if not isinstance(value, str) or not value.strip():
        raise SourceManifestError(f"sources[{index}].{key} must be a non-empty string")
    value = value.strip()
    limit = 2_048 if key == "url" else 200
    if len(value) > limit:
        raise SourceManifestError(f"sources[{index}].{key} is too long")
    return value


def validate_source_manifest(payload: Any) -> tuple[CuratedSource, ...]:
    if not isinstance(payload, dict):
        raise SourceManifestError("Source manifest must be a JSON object")
    if payload.get("format") != SOURCE_MANIFEST_FORMAT:
        raise SourceManifestError(f"format must be {SOURCE_MANIFEST_FORMAT!r}")
    if payload.get("version") != SOURCE_MANIFEST_VERSION:
        raise SourceManifestError(
            f"Unsupported source manifest version: {payload.get('version')!r}"
        )
    raw_sources = payload.get("sources")
    if not isinstance(raw_sources, list):
        raise SourceManifestError("sources must be a JSON array")
    if len(raw_sources) > MAX_SOURCES:
        raise SourceManifestError(f"A manifest may contain at most {MAX_SOURCES} sources")

    result: list[CuratedSource] = []
    seen_urls: set[str] = set()
    for index, raw in enumerate(raw_sources):
        if not isinstance(raw, dict):
            raise SourceManifestError(f"sources[{index}] must be a JSON object")
        unexpected = set(raw) - _SOURCE_FIELDS
        if unexpected:
            names = ", ".join(sorted(unexpected))
            raise SourceManifestError(f"sources[{index}] has unsupported fields: {names}")
        title = _required_text(raw, "title", index)
        raw_url = _required_text(raw, "url", index)
        raw_parts = urlsplit(raw_url)
        try:
            raw_port = raw_parts.port
        except ValueError as error:
            raise SourceManifestError(f"sources[{index}].url has an invalid port") from error
        if raw_parts.scheme not in {"http", "https"} or not raw_parts.hostname:
            raise SourceManifestError(f"sources[{index}].url must be an HTTP(S) URL")
        if raw_parts.username or raw_parts.password:
            raise SourceManifestError(f"sources[{index}].url must not contain credentials")
        if raw_port is not None and raw_port not in {80, 443}:
            raise SourceManifestError(f"sources[{index}].url must use a standard HTTP(S) port")
        url = canonicalize_url(raw_url)
        category = _required_text(raw, "category", index)
        content_kind = raw.get("content_kind")
        if content_kind not in {"paper", "news"}:
            raise SourceManifestError(
                f"sources[{index}].content_kind must be paper or news"
            )
        enabled = raw.get("enabled", True)
        if not isinstance(enabled, bool):
            raise SourceManifestError(f"sources[{index}].enabled must be a boolean")
        if url in seen_urls:
            raise SourceManifestError(f"Duplicate source URL: {url}")
        seen_urls.add(url)
        result.append(CuratedSource(title, url, category, content_kind, enabled))
    return tuple(result)


def load_source_catalog() -> tuple[CuratedSource, ...]:
    resource = files("herald").joinpath("data/sources.json")
    try:
        payload = json.loads(resource.read_text(encoding="utf-8"))
    except (OSError, json.JSONDecodeError) as error:
        raise SourceManifestError(f"Could not load packaged source catalog: {error}") from error
    return validate_source_manifest(payload)


def create_source_manifest(sources: Iterable[Mapping[str, Any]]) -> dict[str, Any]:
    portable = [
        {
            "title": source["title"],
            "url": source["url"],
            "category": source["category"],
            "content_kind": source.get("content_kind", "paper"),
            "enabled": bool(source.get("enabled", True)),
        }
        for source in sources
    ]
    portable.sort(
        key=lambda source: (
            str(source["content_kind"]),
            str(source["category"]).casefold(),
            str(source["title"]).casefold(),
            str(source["url"]),
        )
    )
    validated = validate_source_manifest(
        {"format": SOURCE_MANIFEST_FORMAT, "version": SOURCE_MANIFEST_VERSION, "sources": portable}
    )
    return {
        "format": SOURCE_MANIFEST_FORMAT,
        "version": SOURCE_MANIFEST_VERSION,
        "sources": [source.to_dict() for source in validated],
    }


CURATED_SOURCES = load_source_catalog()
RESEARCH_SOURCES = tuple(source for source in CURATED_SOURCES if source.content_kind == "paper")
NEWS_SOURCES = tuple(source for source in CURATED_SOURCES if source.content_kind == "news")

File: internal/sources/sources.go
// Package sources handles Herald's portable subscription format and the
// versioned catalog of public feeds shipped with the application.
//
// A manifest carries subscriptions only. It deliberately excludes articles,
// reading history, rankings, refresh metadata, and database identifiers, so an
// export can be shared or moved between machines without leaking personal data.
package sources

import (
	"embed"
	"encoding/json"
	"fmt"
	"sort"
	"strings"

	"github.com/PabloMoralesEscandon/herald/internal/feed"
	"github.com/PabloMoralesEscandon/herald/internal/urlx"
)

// Manifest identity and limits.
const (
	ManifestFormat  = "herald.sources"
	ManifestVersion = 1
	MaxSources      = 1_000
	maxTitleLength  = 200
	maxURLLength    = 2_048
	// MaxManifestBytes bounds an imported document before it is parsed.
	MaxManifestBytes = 1_000_000
)

//go:embed data/sources.json
var catalogFS embed.FS

// ManifestError reports an invalid or unsafe manifest. Every message names the
// offending index so a hand-edited file can be corrected.
type ManifestError struct{ Reason string }

func (e *ManifestError) Error() string { return e.Reason }

func manifestErrorf(format string, args ...any) *ManifestError {
	return &ManifestError{Reason: fmt.Sprintf(format, args...)}
}

// Source is one portable subscription.
type Source struct {
	Title       string `json:"title"`
	URL         string `json:"url"`
	Category    string `json:"category"`
	ContentKind string `json:"content_kind"`
	Enabled     bool   `json:"enabled"`
}

// Manifest is the document exchanged by export and import.
type Manifest struct {
	Format  string   `json:"format"`
	Version int      `json:"version"`
	Sources []Source `json:"sources"`
}

var allowedFields = map[string]bool{
	"title": true, "url": true, "category": true,
	"content_kind": true, "enabled": true,
}

// ValidateManifest checks a decoded manifest completely before any of it is
// applied, so an import is all-or-nothing.
func ValidateManifest(payload []byte) ([]Source, error) {
	var envelope map[string]json.RawMessage
	if err := json.Unmarshal(payload, &envelope); err != nil {
		return nil, &ManifestError{Reason: "Source manifest must be a JSON object"}
	}
	var format string
	if err := json.Unmarshal(envelope["format"], &format); err != nil || format != ManifestFormat {
		return nil, manifestErrorf("format must be %q", ManifestFormat)
	}
	var version any
	_ = json.Unmarshal(envelope["version"], &version)
	number, ok := version.(float64)
	if !ok || int(number) != ManifestVersion || number != float64(int(number)) {
		return nil, manifestErrorf("Unsupported source manifest version: %v", version)
	}

	var rawSources []map[string]any
	if err := json.Unmarshal(envelope["sources"], &rawSources); err != nil {
		return nil, &ManifestError{Reason: "sources must be a JSON array"}
	}
	if len(rawSources) > MaxSources {
		return nil, manifestErrorf("A manifest may contain at most %d sources", MaxSources)
	}

	result := make([]Source, 0, len(rawSources))
	seen := make(map[string]bool, len(rawSources))
	for index, raw := range rawSources {
		source, err := validateSource(raw, index)
		if err != nil {
			return nil, err
		}
		if seen[source.URL] {
			return nil, manifestErrorf("Duplicate source URL: %s", source.URL)
		}
		seen[source.URL] = true
		result = append(result, source)
	}
	return result, nil
}

func validateSource(raw map[string]any, index int) (Source, error) {
	var unexpected []string
	for field := range raw {
		if !allowedFields[field] {
			unexpected = append(unexpected, field)
		}
	}
	if len(unexpected) > 0 {
		sort.Strings(unexpected)
		return Source{}, manifestErrorf("sources[%d] has unsupported fields: %s",
			index, strings.Join(unexpected, ", "))
	}

	title, err := requiredText(raw, "title", index)
	if err != nil {
		return Source{}, err
	}
	rawURL, err := requiredText(raw, "url", index)
	if err != nil {
		return Source{}, err
	}
	if err := validateSourceURL(rawURL, index); err != nil {
		return Source{}, err
	}
	canonical, canonErr := feed.CanonicalizeURL(rawURL)
	if canonErr != nil {
		return Source{}, manifestErrorf("sources[%d].url has an invalid port", index)
	}
	category, err := requiredText(raw, "category", index)
	if err != nil {
		return Source{}, err
	}
	kind, _ := raw["content_kind"].(string)
	if kind != "paper" && kind != "news" {
		return Source{}, manifestErrorf("sources[%d].content_kind must be paper or news", index)
	}
	enabled := true
	if value, present := raw["enabled"]; present {
		flag, ok := value.(bool)
		if !ok {
			return Source{}, manifestErrorf("sources[%d].enabled must be a boolean", index)
		}
		enabled = flag
	}
	return Source{
		Title:       title,
		URL:         canonical,
		Category:    category,
		ContentKind: kind,
		Enabled:     enabled,
	}, nil
}

// validateSourceURL rejects URLs Herald must never fetch: non-web schemes,
// embedded credentials, and non-standard ports.
func validateSourceURL(rawURL string, index int) error {
	parts := urlx.Split(rawURL)
	port, present, err := parts.Port()
	if err != nil {
		return manifestErrorf("sources[%d].url has an invalid port", index)
	}
	if parts.Scheme != "http" && parts.Scheme != "https" || parts.Hostname() == "" {
		return manifestErrorf("sources[%d].url must be an HTTP(S) URL", index)
	}
	if parts.HasUserinfo() {
		return manifestErrorf("sources[%d].url must not contain credentials", index)
	}
	if present && port != 80 && port != 443 {
		return manifestErrorf("sources[%d].url must use a standard HTTP(S) port", index)
	}
	return nil
}

func requiredText(raw map[string]any, key string, index int) (string, error) {
	value, ok := raw[key].(string)
	if !ok || strings.TrimSpace(value) == "" {
		return "", manifestErrorf("sources[%d].%s must be a non-empty string", index, key)
	}
	value = strings.TrimSpace(value)
	limit := maxTitleLength
	if key == "url" {
		limit = maxURLLength
	}
	if len([]rune(value)) > limit {
		return "", manifestErrorf("sources[%d].%s is too long", index, key)
	}
	return value, nil
}

// CreateManifest builds a portable document from stored source rows, sorted so
// that repeated exports of unchanged subscriptions are byte-identical.
func CreateManifest(rows []Source) (Manifest, error) {
	portable := append([]Source{}, rows...)
	sort.SliceStable(portable, func(left, right int) bool {
		a, b := portable[left], portable[right]
		if a.ContentKind != b.ContentKind {
			return a.ContentKind < b.ContentKind
		}
		if !strings.EqualFold(a.Category, b.Category) {
			return strings.ToLower(a.Category) < strings.ToLower(b.Category)
		}
		if !strings.EqualFold(a.Title, b.Title) {
			return strings.ToLower(a.Title) < strings.ToLower(b.Title)
		}
		return a.URL < b.URL
	})
	document, err := json.Marshal(Manifest{
		Format:  ManifestFormat,
		Version: ManifestVersion,
		Sources: portable,
	})
	if err != nil {
		return Manifest{}, err
	}
	// Round-trip through validation so an export can never emit a document
	// Herald would refuse to import.
	validated, err := ValidateManifest(document)
	if err != nil {
		return Manifest{}, err
	}
	return Manifest{Format: ManifestFormat, Version: ManifestVersion, Sources: validated}, nil
}

// Catalog returns the versioned public feeds shipped with the application.
//
// It is embedded in the binary, so `herald init` works from any directory with
// no data files alongside the executable.
func Catalog() ([]Source, error) {
	document, err := catalogFS.ReadFile("data/sources.json")
	if err != nil {
		return nil, manifestErrorf("Could not load packaged source catalog: %v", err)
	}
	catalog, err := ValidateManifest(document)
	if err != nil {
		return nil, manifestErrorf("Could not load packaged source catalog: %v", err)
	}
	return catalog, nil
}
